File: ielts-vocab-master/Python/Flask/auth/firebase_login.py
from flask import Blueprint, jsonify, request, session

from db import get_db
from oauth_common import firebase_verify_id_token, get_firebase_api_key, resolve_oauth_user

firebase_login_bp = Blueprint('firebase_login', __name__)


@firebase_login_bp.route('/firebase_login', methods=['POST'])
def firebase_login():
    body     = request.get_json(silent=True) or {}
    id_token = body.get('idToken', '')

    if not isinstance(id_token, str) or not id_token:
        return jsonify({'error': 'Missing sign-in token.'}), 400

    api_key = get_firebase_api_key()
    if not api_key:
        return jsonify({'error': 'Social sign-in is not configured yet.'}), 500

    profile = firebase_verify_id_token(id_token, api_key)
    if not profile:
        return jsonify({'error': 'Could not verify sign-in. Please try again.'}), 401

    conn    = get_db()
    user_id = resolve_oauth_user(conn, profile)

    conn.execute("UPDATE users SET last_login = datetime('now') WHERE id = ?", (user_id,))
    conn.commit()

    session.clear()
    session['user_id'] = user_id
    return jsonify({'ok': True})

File: ielts-vocab-master/PHP/auth/firebase_login.php
<?php
session_start();
header('Content-Type: application/json');
require __DIR__ . '/db.php';
require __DIR__ . '/oauth_common.php';

$input = json_decode(file_get_contents('php://input'), true) ?? [];
$idToken = is_string($input['idToken'] ?? null) ? $input['idToken'] : '';

if ($idToken === '') {
    http_response_code(400);
    echo json_encode(['error' => 'Missing sign-in token.']);
    exit;
}

$apiKey = get_firebase_api_key();
if (!$apiKey) {
    http_response_code(500);
    echo json_encode(['error' => 'Social sign-in is not configured yet.']);
    exit;
}

$profile = firebase_verify_id_token($idToken, $apiKey);
if (!$profile) {
    http_response_code(401);
    echo json_encode(['error' => 'Could not verify sign-in. Please try again.']);
    exit;
}

$pdo = get_db();
$userId = resolve_oauth_user($pdo, $profile);

$pdo->prepare("UPDATE users SET last_login = datetime('now') WHERE id = ?")->execute([$userId]);

session_regenerate_id(true);
$_SESSION['user_id'] = $userId;

echo json_encode(['ok' => true]);
